src/icp_canister_backend/src/lib.rs

```diff
@@ -245,20 +282,23 @@ pub async fn withdraw(deposit_id: u64) -> Result<(), types::PoolError> {
         return Err(types::PoolError::NotOwner);
     }
 
-    if now < deposit.unlocktime {
+    if deposit.episode >= current_episode {
         return Err(types::PoolError::TimelockNotExpired);
     }
 
-    let withdrawal_amount = POOL_STATE.with(|state| {
-        let pool_state = state.borrow().get().clone();
+    let episode_data = EPISODES.with(|episodes| {
+        episodes
+            .borrow()
+            .get(&deposit.episode)
+            .ok_or(types::PoolError::NoDeposit)
+    })?;
 
-        if pool_state.total_shares == Nat::from(0u128) {
-            deposit.shares.clone()
-        } else {
-            deposit.shares.clone() * pool_state.total_assets.clone()
-                / pool_state.total_shares.clone()
-        }
-    });
+    let withdrawal_amount = if episode_data.episode_shares == Nat::from(0u64) {
```

Comment:
There should never be a case when episode which is being unstake has zero shares.